language-support/ts/daml-ledger/index.ts

```diff
@@ -329,6 +331,277 @@ type LedgerOptions = {
   reconnectThreshold?: number;
 }
 
+class StreamEventEmitter<T extends object, K, I extends string, State> extends EventEmitter implements Stream<T, K, I, State> {
+
+    /** Passed at construction time, called _before_ the 'close' event is emitted */
+    private readonly beforeClosing: () => void;
+
+    constructor({beforeClosing}: { beforeClosing: () => void }) {
+        super();
+        this.beforeClosing = beforeClosing;
+    }
+
+    close(): void {
+        this.beforeClosing();
+        this.emit('close', { code: 4000, reason: "called .close()" });
+        this.removeAllListeners();
+    }
+
+}
+
+type QueryResponseStream<T extends object, K, I extends string> = StreamEventEmitter<T, K, I, readonly CreateEvent<T, K, I>[]>;
+
+type StreamingQuery<T extends object, K, I extends string> = {
+    template: Template<T, K, I>,
+    queries: Query<T>[],
+    stream: QueryResponseStream<T, K, I>,
+    offset?: string | null, // undefined -> nothing has been received, null -> the JSON API returned a null offset
```

Comment:
My TypeScript is a bit rusty; would it be possible to define some sort of enum or sentinel objects here? Googling yielded [this](https://2ality.com/2020/01/special-values-typescript.html), which suggests something like:

```typescript
const NoOffsetReceivedYet = Symbol('NoOffsetReceivedYet');
const NullOffsetReceived = Symbol('NullOffsetReceived');
...
    offset: string | typeof NoOffsetReceivedYet | typeof NullOffsetReceived,
```

Reply:
I think this sounds very reasonable, thanks.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/10221/commits/e87bc5ca21b3d8f8b7b23ac6dff5fc40fe33e8f0